### ISSUE TITLE
Webcam video does not appear, with no error messages, when using videojs-record with VueJS

### ISSUE THREAD

**Issue by hmexx** [collab-project/videojs-record#212]
The video component looks correct, but when you click it, and after you've given it video permission (chrome, firefox) the area where the video would show up stays black. The record/stop buttons do not show up like they should. There are no errors in the log. The debug output from initialising video.js is:

> VIDEOJS: Using video.js 6.6.3 with videojs-record 2.1.0 and recordrtc 5.4.6

Here is the single-file vuejs component code. I've tried to keep it to a minimum. Hope someone has an idea because I can't get it to work.

Thanks.

```javascript
<template>
  <div data-vjs-player>
    <video id="myVideo"
           class="video-js vjs-default-skin"></video>
  </div>
</template>

<script lang="ts">
import * as V from "vue-property-decorator";
import * as videojs from "video.js";
import * as recordrtc from "recordrtc";
import "webrtc-adapter";
import "videojs-record/dist/css/videojs.record.css";
import "videojs-record/dist/videojs.record.js";

@V.Component({
  name: "cvideorecorder"
})
export default class cVideoRecorder extends V.Vue {
  player: videojs.Player | null = null;

  beforeDestroy() {
    if (!this.player || !this.player!.dispose) return;
    this.player!.dispose();
  }

  mounted() {
    this.player = videojs(
      "myVideo",
      {
        controls: true,fluid: false,
        width: 320, height: 240,
        plugins: {
          record: {
            audio: true, video: true,
            maxLength: 10, debug: true
          }
        }
      },
      function() {
        let v = videojs as any;
        v.log("Using video.js", v.VERSION,"with videojs-record",
        v.getPluginVersion("record"), "and recordrtc", recordrtc.version
        );
      }
    );

    this.player.on("error", function(error: any) {
      console.log(error);
    });
  }
}
</script>
<style scoped></style>

```

**Comment by thijstriemstra** [comment]
@hmexx I assume you cannot re-create this with any of the demo's at https://github.com/collab-project/videojs-record#examples and it's only happening in your VueJS app?

**Comment by thijstriemstra** [comment]
@hmexx I see you're importing the following (seemingly without a purpose), perhaps these screw up something:

```javascript
import "videojs-contrib-hls";
import "videojs-flash";
```

**Comment by hmexx** [comment]
Hi @thijstriemstra 

You are right. Those are unnecessary and I've just removed them, with no positive or negative change. I was adding related imports in my desperation to make it work.

The demo's you linked to work fine on my set-up. I've also looked at some React code to see if they do anything special and there's nothing I could find that I've not already taken care of.  I could not find a single mention of VueJS combined with videojs-record on the web.

Thanks for any ideas you might have.

**Comment by hmexx** [comment]
@thijstriemstra I'd get to the bottom of this myself if I had an error message to go on. Quite weird how it's a silent failure. Is there anything I can do on my end to help you point me in the right direction? I have a publicly accessible page with the malfunction component up and running. Would it be useful if I sent you that to inspect the page?

I couldn't find anything weird in there myself.

Thanks!

**Comment by thijstriemstra** [comment]
@hmexx sure, I'll take a look.

**Comment by hmexx** [comment]
@thijstriemstra 

Great! I'll leave this up for a few days

http://bit.ly/2GwQ4HP

Oddly enough, I've now noticed that I've moved it to a non-localhost domain, that it does not actually ask for video permission. On my dev environment, the browser already has camera permission, and you can see the camera light go on, but nothing shows up (black video). Maybe the difference between the two can help understand what's going wrong?

Thanks so much.

**Comment by hmexx** [comment]
I've managed to get some kind of error to appear by changing my set-up. Perhaps this could help the investigation. I had the following in my webpack.conf set-up.

```javascript
 resolve: {
    extensions: ['.ts', '.js', '.vue', '.json'],
    alias: {
      'vue$': 'vue/dist/vue.esm.js',
      '@': resolve('src'),
      'video.js$': 'video.js/dist/video.cjs.js',
      'videojs-contrib-hls': 'videojs-contrib-hls/dist/videojs-contrib-hls'
    }
  },
  plugins: [
    new webpack.ProvidePlugin({
      videojs: 'video.js',
      'window.videojs': 'video.js',
    }),
    new webpack.DefinePlugin({
      'typeof global': JSON.stringify('undefined'),
    }),
  ],
```

if I replace `video.cjs.js` with `video.es.js` I get the following error in the console:

`videojs.getComponent is not a function`

if I replace `video.cjs.js` with `video.min.js` then it goes back to the original behavious (silently failing)

**Comment by hmexx** [comment]
Another potential hint. If you open the webpage I've linked to in Microsoft Edge, then the behaviour is slightly different to chrome. The record button DOES show up after you click the main button. But it's still not functional in any way.

**Comment by hmexx** [comment]
@thijstriemstra 

Wait a minute, the react example from your repo does not work either?!

https://collab-project.github.io/videojs-record/examples/react/index.html

The exact same behaviour! This gives me some hope that I'm not going crazy. Let me know if you're seeing the same thing on your end.

**Comment by hmexx** [comment]
Well, I solved it. About 8 hours of pain and suffering … One of those dev days that makes you want to change jobs, lol.

OK, for future reference, and I hope this helps someone, the exact webpack config that is needed is below. What was missing was a line for RecordRTC in there.

```javascript
  resolve: {
    extensions: ['.ts', '.js', '.vue', '.json'],
    alias: {
      'vue$': 'vue/dist/vue.esm.js',
      '@': resolve('src')
    }
  },
  plugins: [
    new webpack.ProvidePlugin({
      videojs: 'video.js',
      'window.videojs': 'video.js',
      RecordRTC: 'recordrtc', 'window.RecordRTC': 'recordrtc'
    })
  ],
```

combine this with the code in my OP and you have a working VueJS video recording component.

@thijstriemstra great plugin btw. keep up the good work.

**Comment by thijstriemstra** [comment]
@hmexx you never mentioned you were using webpack... Glad you solved it (react component works afterall I guess, you removed the comment).

**Comment by hmexx** [comment]
Yes the react one works fine. I removed the comment sneakily... I did not want to admit that I forgot plug my webcam back in when I tried the react version!

And you're right, mentioning webpack would have been useful, although I think it would have still been a struggle. I don't think the exact config required, as detailed above, is available all in one place on the web. It's there now though :)

Have a good day.
